sphinx/builders/__init__.py

```diff
@@ -420,8 +420,18 @@ def read(self) -> list[str]:
             self._read_serial(docnames)
 
         if self.config.root_doc not in self.env.all_docs:
-            raise SphinxError('root file %s not found' %
-                              self.env.doc2path(self.config.root_doc))
+            if ("**" in self.config.exclude_patterns or
```

Comment:
This fails if you have an empty `exclude_patterns` but `include_patterns = ['a.rst', 'b.rst']` only. 

Instead, check whether `self.config.root_doc` is matched by `include_patterns`. Since the include/exclude patterns follow fnmatch-style patterns, you can check whether the root doc is matched (or not) by those patterns using [fnmatch.fnmatch](https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch).

A similar logic can be applied for `exclude_patterns` where you check whether `self.config.root_doc` is *not* matched by `exclude_patterns`. You might also want to have a look at [source_suffix](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-source_suffix)

Reply:
Yes, it can be configured with [root_doc](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-root_doc)

Reply:
I tested the following
Case A
root file name = index.rst
root_doc = default
`exclude_patterns=[] `and `include_patterns = ['a.rst', 'b.rst']`
output  an error message saying `index.rst` not in `include_patterns`

Case B
root file name = index.rst
root_doc = default
`exclude_patterns=[]` and `include_patterns = ['index.rst', 'b.rst']`
no error, build successfully

Case C
root file name = a.rst
root_doc = default
`exclude_patterns=[]` and `include_patterns = ['a.rst', 'b.rst']`
output  an error message saying `index.rst` not in `include_patterns`

Case D
root file name = a.rst
root_doc = 'a'
`exclude_patterns=[]` and `include_patterns = ['a.rst', 'b.rst']`
no error, build successfully

Are these cases intended behavior, or which one should I modify the error message?

Reply:
Can you also check with the `source_suffix` configuration where you don't use the .rst extension but use .txt extensions?

Reply:
(Ah and sorry for not answering question, but they are correct).

You should also test with some exclusion patterns and inclusion patterns where both contain the index (I don't know which one has the priority)

Reply:
No worries. I got your point that I need to take all souce_suffix (.rst, .txt, .md) into account.
From the documentation, the exclude_patterns has a higher priority.